drivers/acpi/acpi.c

```diff
@@ -0,0 +1,543 @@
+	if (ACPI_FAILURE(status)) {
+		ACPI_EXCEPTION((AE_INFO, status, "While enabling ACPI"));
+		goto exit;
+	}
+
+	/* Install local handlers */
+	status = install_handlers();
+	if (ACPI_FAILURE(status)) {
+		ACPI_EXCEPTION((AE_INFO, status, "While installing handlers"));
+		goto exit;
+	}
+
+	/* Create the ACPI namespace from ACPI tables */
+	status = AcpiLoadTables();
+	if (ACPI_FAILURE(status)) {
+		ACPI_EXCEPTION((AE_INFO, status, "While loading ACPI tables"));
+		goto exit;
+	}
+
+	/* Complete the ACPI namespace object initialization */
+	status = AcpiInitializeObjects(ACPI_FULL_INITIALIZATION);
+	if (ACPI_FAILURE(status)) {
+		ACPI_EXCEPTION((AE_INFO, status, "While initializing ACPI objects"));
+	}
+exit:
+
+	return status;
+}
+
+static ACPI_NAMESPACE_NODE *acpi_name_Lookup(char *name)
+{
+	char *path;
+	ACPI_STATUS status;
+	ACPI_NAMESPACE_NODE *node = NULL;
+
+	LOG_DBG("");
+
+	status = AcpiNsInternalizeName(name, &path);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("Invalid namestring: %s", name);
+		return NULL;
+	}
+
+	status = AcpiNsLookup(NULL, path, ACPI_TYPE_ANY, ACPI_IMODE_EXECUTE,
+			      ACPI_NS_NO_UPSEARCH | ACPI_NS_DONT_OPEN_SCOPE, NULL, &node);
+	if (ACPI_FAILURE(status)) {
+		LOG_ERR("Could not locate name: %s, %d", name, status);
+	}
+
+	ACPI_FREE(path);
+	return node;
+}
+
+static ACPI_NAMESPACE_NODE *acpi_evaluate_method(char *bus_name, char *method)
+{
+	ACPI_NAMESPACE_NODE *node;
+	ACPI_NAMESPACE_NODE *handle;
+	ACPI_NAMESPACE_NODE *prt_node = NULL;
+
+	LOG_DBG("%s", bus_name);
+
+	handle = acpi_name_Lookup(bus_name);
+	if (!handle) {
+		LOG_ERR("No ACPI node with given name: %s", bus_name);
+		goto exit;
+	}
+
+	if (handle->Type != ACPI_TYPE_DEVICE) {
+		LOG_ERR("No ACPI node foud with given name: %s", bus_name);
+		goto exit;
+	}
+
+	node = ACPI_CAST_PTR(ACPI_NAMESPACE_NODE, handle);
+
+	(void)AcpiGetHandle(node, method, ACPI_CAST_PTR(ACPI_HANDLE, &prt_node));
+
+	if (!prt_node) {
+		LOG_ERR("No entry for the ACPI node with given name: %s", bus_name);
+		goto exit;
+	}
+	return node;
+exit:
+	return NULL;
+}
+
+static ACPI_STATUS acpi_enable_pic_mode(void)
+{
+	ACPI_STATUS status;
+	ACPI_OBJECT_LIST arg_list;
+	ACPI_OBJECT arg[1];
+
+	arg_list.Count = 1;
+	arg_list.Pointer = arg;
+
+	arg[0].Type = ACPI_TYPE_INTEGER;
+	arg[0].Integer.Value = 1;
+
+	status = AcpiEvaluateObject(NULL, "\\_PIC", &arg_list, NULL);
+	if (ACPI_FAILURE(status)) {
+		LOG_WRN("error While executing \\_pic method: %d", status);
+	}
+
+	return status;
+}
+
+static int acpi_get_irq_table(char *bus_name, ACPI_PCI_ROUTING_TABLE *rt_table, uint32_t rt_size)
+{
+	ACPI_BUFFER rt_buffer;
+	ACPI_NAMESPACE_NODE *node;
+	int status;
+
+	LOG_DBG("%s", bus_name);
+
+	node = acpi_evaluate_method(bus_name, METHOD_NAME__PRT);
+	if (!node) {
+		LOG_ERR("Evaluation failed for given device: %s", bus_name);
+		return -ENODEV;
+	}
+
+	rt_buffer.Pointer = rt_table;
+	rt_buffer.Length = ACPI_DEBUG_BUFFER_SIZE;
+
+	status = AcpiGetIrqRoutingTable(node, &rt_buffer);
```

Comment:
This is for retrieve PCH interrupt routing information for PCI devices using _PRT method. We will check IRQ information provided in PCI config space and if not present, then try to retrieve via ACPI and configure PCI interrupt ctrl register. Currently if CONFIG_ACPI enabled then it will not invoke arch_irq_allocate() (please see pcie.c). But looks we need to mark IRQ numbers retrieved via PRT as allocated/reserved to avoid conflict if someone try to allocated arch_irq_allocate() such as for MSI. I will try to figure out this and update the PR.

Reply:
updated code to mark PRT IRQs as reserved using arch_irq_set_used (please see line number 219).